cpp/include/ray/api/ray_register.h

```diff
@@ -0,0 +1,201 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <ray/api/serializer.h>
+#include "absl/utility/utility.h"
+
+#include <boost/callable_traits.hpp>
+#include <functional>
+#include <string>
+#include <unordered_map>
+
+namespace ray {
+namespace api {
+
+enum ErrorCode {
+  OK = 0,
+  FAIL = 1,
+};
+
+struct VoidResponse {
+  int error_code;
+  std::string error_msg;
+
+  MSGPACK_DEFINE(error_code, error_msg);
+};
+
+template <typename T>
+struct Response {
+  int error_code;
+  std::string error_msg;
+  T data;
+
+  MSGPACK_DEFINE(error_code, error_msg, data);
+};
+
+template <class, class>
+struct AddType;
+
+template <class First, class... Second>
+struct AddType<First, std::tuple<Second...>> {
+  using type = std::tuple<First, Second...>;
+};
+
+template <class First, class Second>
+using AddType_t = typename AddType<First, Second>::type;
+
+template <typename OUT, typename IN>
+inline static OUT AddressOf(IN in) {
+  union {
+    IN in;
+    OUT out;
+  } u = {in};
+
+  return u.out;
+};
+
+class Router {
+ public:
+  static Router &Instance() {
+    static Router instance;
+    return instance;
+  }
+
+  template <typename Function>
+  bool RegisterHandler(std::string const &name, const Function &f) {
+    auto pair = func_ptr_to_key_map_.emplace(AddressOf<uintptr_t>(f), name);
+    if (!pair.second) {
+      return false;
+    }
+
+    RegisterNonmemberFunc(name, f);
+    return true;
+  }
+
+  std::string Route(const char *data, std::size_t size) {
+    std::string result;
+    do {
+      try {
+        auto p = Serializer::Deserialize<std::tuple<std::string>>(data, size);
+        auto &func_name = std::get<0>(p);
+        auto it = map_invokers_.find(func_name);
+        if (it == map_invokers_.end()) {
+          result = PackArgsStr(ErrorCode::FAIL, "unknown function: " + func_name);
+          break;
+        }
+
+        it->second(data, size, result);
+        if (result.size() >= MAX_BUF_LEN) {
+          result = PackArgsStr(
+              ErrorCode::FAIL,
+              "the response result is out of range: more than 10M " + func_name);
+        }
+      } catch (const std::exception &ex) {
+        result = PackArgsStr(ErrorCode::FAIL, ex.what());
+      }
+    } while (0);
+
+    return result;
+  }
+
+ private:
+  Router() = default;
+  ~Router() = default;
+  Router(const Router &) = delete;
+  Router(Router &&) = delete;
+
+  template <typename F, size_t... I, typename Arg, typename... Args>
+  static typename std::result_of<F(Args...)>::type CallHelper(
+      const F &f, const absl::index_sequence<I...> &, std::tuple<Arg, Args...> tup) {
+    return f(std::move(std::get<I + 1>(tup))...);
+  }
+
+  template <typename F, typename Arg, typename... Args>
+  static typename std::enable_if<
+      std::is_void<typename std::result_of<F(Args...)>::type>::value>::type
+  Call(const F &f, std::string &result, std::tuple<Arg, Args...> tp) {
```

Comment:
Can we split the function management and invoking? In Java, we have `FunctionManager` and `TaskExecutor`. In Python, we have "function_manager.py" and `task_execution_handler`. It also reminds me that better to rename `Router` to `FunctionManager`.

Reply:
I think the name "Router" is ok, as mentioned [here](https://github.com/ray-project/ray/pull/14436#discussion_r585475698)

Reply:
You are creating a new term for Ray, but the functionality is not new. It's already implemented in Python and Java. Why not reuse existing terms?

Reply:
Have added function_manager and task_execution_handler.